std/process.d

```diff
@@ -1050,54 +1050,107 @@ private Pid spawnProcessPosix(scope const(char[])[] args,
 
                         // Get the maximum number of file descriptors that could be open.
                         rlimit r;
                         if (getrlimit(RLIMIT_NOFILE, &r) != 0)
-                        {
                             abortOnError(forkPipeOut, InternalError.getrlimit, .errno);
-                        }
+
                         immutable maxDescriptors = cast(int) r.rlim_cur;
 
-                        immutable maxToClose = maxDescriptors - lowfd;
+                        // Missing druntime declaration
+                        pragma(mangle, "dirfd")
+                        extern(C) nothrow @nogc int dirfd(DIR* dir);
 
-                        // Call poll() to see which ones are actually open:
-                        auto pfds = cast(pollfd*) malloc(pollfd.sizeof * maxToClose);
-                        if (pfds is null)
-                        {
-                            abortOnError(forkPipeOut, InternalError.malloc, .errno);
-                        }
-                        foreach (i; 0 .. maxToClose)
+                        DIR* dir = null;
+
+                        // We read from /dev/fd or /proc/self/fd only if the limit is high enough
+                        if (maxDescriptors > 128*1024)
                         {
-                            pfds[i].fd = i + lowfd;
-                            pfds[i].events = 0;
-                            pfds[i].revents = 0;
+                            // Try to open the directory /dev/fd or /proc/self/fd
+                            dir = opendir("/dev/fd");
+                            if (dir is null) dir = opendir("/proc/self/fd");
                         }
-                        if (poll(pfds, maxToClose, 0) >= 0)
+
+                        // If we have a directory, close all file descriptors except stdin, stdout, stderr and forkPipeOut
+                        if (dir)
                         {
-                            foreach (i; 0 .. maxToClose)
+                            scope(exit) closedir(dir);
+
+                            int opendirfd = dirfd(dir);
+
+                            // Iterate over all file descriptors
+                            while (true)
                             {
-                                // POLLNVAL will be set if the file descriptor is invalid.
-                                if (!(pfds[i].revents & POLLNVAL)) close(pfds[i].fd);
+                                dirent* entry = readdir(dir);
+
+                                if (entry is null) break;
+                                if (entry.d_name[0] == '.') continue;
+
+                                int fd = atoi(cast(char*) entry.d_name);
+
+                                // Don't close stdin, stdout, stderr, forkPipeOut or the directory file descriptor
+                                if (fd < 3 || fd == forkPipeOut || fd == opendirfd) continue;
```

Comment:
Should this `3` should be `lowfd`?

Reply:
Yep, missed that one.

Reply:
Though, technically, lowfd is always 3 ;)

Reply:
Oh wait, no, it's 4. Weird.

Reply:
ugh, there's a lot of use of forkpipeout here too. Need to massage this a bit.

Reply:
ok, I think it's better.